### PR TITLE
[enhancement](Nereids) avoiding broadcast join heuristically and pruning more in CostAndEnforceJob

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes

1. When the rowCount exceeds a certain threshold, refrain from generating a broadcast join.
2. Only enforce the best expression in CostAndEnforce Job, rather than enforcing every expression.
3. Remove lower bound group pruning

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java [hunk 1]
```diff
@@ -106,6 +106,10 @@ public DistributionSpec getDistributionSpec() {
         return distributionSpec;
     }
 
+    public boolean isDistributorProperties() {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java:109]: maybe `distributionOnlyProperties` is a better name

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java [hunk 2]
```diff
@@ -165,8 +165,14 @@ public Void visitPhysicalHashJoin(PhysicalHashJoin<? extends Plan, ? extends Pla
...
         }
+
         // for broadcast join
-        if (JoinUtils.couldBroadcast(hashJoin)) {
+        double memLimit = ConnectContext.get().getSessionVariable().getMaxExecMemByte();
+        double rowsLimit = ConnectContext.get().getSessionVariable().getBroadcastRowCountLimit();
+        double brMemlimit = ConnectContext.get().getSessionVariable().getBroadcastHashtableMemLimitPercentage();
+        double datasize = hashJoin.getGroupExpression().get().child(1).getStatistics().computeSize();
+        double rowCount = hashJoin.getGroupExpression().get().child(1).getStatistics().getRowCount();
+        if (JoinUtils.couldBroadcast(hashJoin) && rowCount <= rowsLimit && datasize <= memLimit * brMemlimit) {
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:175]: please remove memcheck and rowcount check from CostModelV1.visitPhysicalDistribute()

Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:175]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java (hunk 1 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java]: done

**Comment by XieJiann** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**